### ISSUE TITLE
Implement LX allocation algorithm based on Imanishi & Xu (2024)

### ISSUE THREAD

**Issue by postmath** [torch-spyre/torch-spyre#2076]
### What feature would you like to be added?

The paper [1] introduces an algorithm for memory allocation based on simulated annealing. It looks like an algorithm that would be suitable for doing LX allocation, with some adjustments. For example, like most papers in this space, as written the algorithm allocates *all* buffers in *as little space as possible*. For the LX allocation application, we want to allocate *as many buffers as possible* (weighted by some metric) in a *fixed* amount of space. However, we thinks this can easily be overcome.

[1]: Imanishi, A. and Xu, Z., 2024, June. A heuristic for periodic memory allocation with little fragmentation to train neural networks. In Proceedings of the 2024 ACM SIGPLAN International Symposium on Memory Management (pp. 82-94).

### Why is this feature needed?

We need an LX allocator that optimizes for the graph as a whole, not in a greedy way.

### Any alternative solution that you may want to share?

There are many other options. One promising one is to use an ILP solver from an existing library, such as [OR-tools](https://developers.google.com/optimization). One advantage of this algorithm is that easily be implemented in standalone code, without extra libraries. 

### Anything else we need to know?

We have a prototype in about 800 lines of Python code. So far, this prototype is essentially a straight implementation of the paper as written; it would need to be adjusted to operate with a fixed amount of space, as discussed above. Another feature that is still needed is dealing with in-place operators (with potentially multiple sources and/or targets for a buffer).

As with any simulated annealing setup, there is a fair number of parameters that can be adjusted, including ones related to the cooling schedule (such as the number of iterations), the initial state, and the number of restarts. We would need to make sure to test the algorithm on a wide variety of FX graphs to ensure the defaults are reasonable. Perhaps we should add a single parameter that would allow the user to adjust how much compute they are willing to invest vs. how important a high quality solution is, with us adjusting the other parameter values accordingly.

As of the opening of this issue, the prototype lives [here](https://github.com/postmath/torch-spyre/tree/imanishi-xu).

The prototype also includes two other fast heuristics that are used as initialization for the initial state of the main solver. These are known as "first fit" and "best fit". They could be offered as faster, lower quality standalone solvers.
